File: Tools/compare_beir_runs.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';

function argsMap(argv) {
  const args = new Map();
  for (let i = 2; i < argv.length; ++i) {
    const key = argv[i];
    if (!key.startsWith('--')) continue;
    if (i + 1 < argv.length && !argv[i + 1].startsWith('--')) args.set(key.slice(2), argv[++i]);
    else args.set(key.slice(2), 'true');
  }
  return args;
}

function required(args, key) {
  const value = args.get(key);
  if (!value) throw new Error(`Missing --${key}`);
  return value;
}

function loadQrels(filePath) {
  const qrels = new Map();
  const lines = fs.readFileSync(filePath, 'utf8').split(/\r?\n/);
  for (const line of lines) {
    if (!line) continue;
    const columns = line.split('\t');
    if (columns.length < 3 || columns[0] === 'query-id') continue;
    const score = Number(columns[2]);
    if (!Number.isFinite(score) || score <= 0) continue;
    if (!qrels.has(columns[0])) qrels.set(columns[0], new Set());
    qrels.get(columns[0]).add(columns[1]);
  }
  return qrels;
}

function loadRun(filePath, maxRank, qrels) {
  const run = new Map();
  const lines = fs.readFileSync(filePath, 'utf8').split(/\r?\n/);
  for (const line of lines) {
    if (!line) continue;
    const parts = line.trim().split(/\s+/);
    if (parts.length < 6) continue;
    const [qid, , docid, rankText, scoreText] = parts;
    const rank = Number(rankText);
    const score = Number(scoreText);
    if (!Number.isFinite(rank) || !Number.isFinite(score) || rank > maxRank) continue;
    if (!qrels.get(qid)?.has(docid)) continue;
    if (!run.has(qid)) run.set(qid, new Map());
    const docs = run.get(qid);
    if (!docs.has(docid)) docs.set(docid, { rank, score });
  }
  return run;
}

function sortedDocs(docSet, runDocs) {
  return [...docSet].sort((a, b) => (runDocs?.get(a)?.rank ?? 1e9) - (runDocs?.get(b)?.rank ?? 1e9));
}

function intersection(a, b) {
  const out = new Set();
  for (const value of a) if (b.has(value)) out.add(value);
  return out;
}

function difference(a, b) {
  const out = new Set();
  for (const value of a) if (!b.has(value)) out.add(value);
  return out;
}

function main() {
  const args = argsMap(process.argv);
  const qrelsPath = required(args, 'qrels');
  const leftPath = required(args, 'left');
  const rightPath = required(args, 'right');
  const leftName = args.get('left-name') ?? 'left';
  const rightName = args.get('right-name') ?? 'right';
  const k = Number(args.get('k') ?? '1000');
  const top = Number(args.get('top') ?? '30');
  const outPrefix = args.get('out-prefix') ?? '';

  const qrels = loadQrels(qrelsPath);
  const left = loadRun(leftPath, k, qrels);
  const right = loadRun(rightPath, k, qrels);
  const qids = [...qrels.keys()].filter(qid => left.has(qid) || right.has(qid)).sort();

  let totalRelevant = 0;
  let totalCommon = 0;
  let totalLeftOnly = 0;
  let totalRightOnly = 0;
  let totalLeftHits = 0;
  let totalRightHits = 0;
  const rows = [];

  for (const qid of qids) {
    const relevant = qrels.get(qid);
    const leftHits = intersection(new Set(left.get(qid)?.keys() ?? []), relevant);
    const rightHits = intersection(new Set(right.get(qid)?.keys() ?? []), relevant);
    const common = intersection(leftHits, rightHits);
    const leftOnly = difference(leftHits, rightHits);
    const rightOnly = difference(rightHits, leftHits);
    const delta = rightHits.size - leftHits.size;

    totalRelevant += relevant.size;
    totalCommon += common.size;
    totalLeftOnly += leftOnly.size;
    totalRightOnly += rightOnly.size;
    totalLeftHits += leftHits.size;
    totalRightHits += rightHits.size;

    rows.push({
      qid,
      relevant: relevant.size,
      [`${leftName}_hits`]: leftHits.size,
      [`${rightName}_hits`]: rightHits.size,
      common: common.size,
      [`${leftName}_only`]: leftOnly.size,
      [`${rightName}_only`]: rightOnly.size,
      delta_right_minus_left: delta,
      [`${leftName}_only_docs`]: sortedDocs(leftOnly, left.get(qid)).slice(0, 10).join(';'),
      [`${rightName}_only_docs`]: sortedDocs(rightOnly, right.get(qid)).slice(0, 10).join(';'),
    });
  }

  console.log(`queries=${qids.length} k=${k} relevant=${totalRelevant}`);
  console.log(`${leftName}_hits=${totalLeftHits} ${rightName}_hits=${totalRightHits} delta=${totalRightHits - totalLeftHits}`);
  console.log(`common=${totalCommon} ${leftName}_only=${totalLeftOnly} ${rightName}_only=${totalRightOnly}`);
  if (totalRelevant) {
    console.log(`recall_${leftName}=${(totalLeftHits / totalRelevant).toFixed(6)} recall_${rightName}=${(totalRightHits / totalRelevant).toFixed(6)}`);
  }

  const worst = [...rows].sort((a, b) => a.delta_right_minus_left - b.delta_right_minus_left).slice(0, top);
  const best = [...rows].sort((a, b) => b.delta_right_minus_left - a.delta_right_minus_left).slice(0, top);
  console.log('\nWorst right-minus-left queries:');
  for (const row of worst) console.log(JSON.stringify(row));
  console.log('\nBest right-minus-left queries:');
  for (const row of best) console.log(JSON.stringify(row));

  if (outPrefix && rows.length) {
    const outputPath = `${outPrefix}.per_query.tsv`;
    const fields = Object.keys(rows[0]);
    const body = [fields.join('\t'), ...rows.map(row => fields.map(field => row[field] ?? '').join('\t'))].join('\n') + '\n';
    fs.mkdirSync(path.dirname(outputPath), { recursive: true });
    fs.writeFileSync(outputPath, body, 'utf8');
    console.log(`wrote ${outputPath}`);
  }
}

main();

File: Tools/compare_beir_runs.py
#!/usr/bin/env python3
import argparse
import csv
from collections import defaultdict
from pathlib import Path


def load_qrels(path: Path):
    qrels = defaultdict(set)
    with path.open("r", encoding="utf-8") as handle:
        reader = csv.reader(handle, delimiter="\t")
        for row in reader:
            if not row or row[0] == "query-id" or len(row) < 3:
                continue
            try:
                score = float(row[2])
            except ValueError:
                continue
            if score > 0:
                qrels[row[0]].add(row[1])
    return qrels


def load_run(path: Path, max_rank: int, qrels):
    run = defaultdict(dict)
    with path.open("r", encoding="utf-8") as handle:
        for line in handle:
            parts = line.strip().split()
            if len(parts) < 6:
                continue
            qid, _, docid, rank_text, score_text, _tag = parts[:6]
            try:
                rank = int(rank_text)
                score = float(score_text)
            except ValueError:
                continue
            if docid not in qrels.get(qid, set()):
                continue
            if rank <= max_rank and docid not in run[qid]:
                run[qid][docid] = (rank, score)
    return run


def sorted_docs(docs, run):
    return sorted(docs, key=lambda docid: run.get(docid, (10**9, 0.0))[0])


def main():
    parser = argparse.ArgumentParser(description="Compare two BEIR/TREC run files against qrels.")
    parser.add_argument("--qrels", required=True)
    parser.add_argument("--left", required=True, help="Usually Lucene run")
    parser.add_argument("--right", required=True, help="Usually MoonShot run")
    parser.add_argument("--left-name", default="left")
    parser.add_argument("--right-name", default="right")
    parser.add_argument("--k", type=int, default=1000)
    parser.add_argument("--top", type=int, default=30, help="Number of worst delta queries to print")
    parser.add_argument("--out-prefix", default="")
    args = parser.parse_args()

    qrels = load_qrels(Path(args.qrels))
    left = load_run(Path(args.left), args.k, qrels)
    right = load_run(Path(args.right), args.k, qrels)
    qids = sorted(qid for qid in qrels if qid in left or qid in right)

    total_relevant = 0
    total_common = 0
    total_left_only = 0
    total_right_only = 0
    total_left_hits = 0
    total_right_hits = 0
    rows = []

    for qid in qids:
        relevant = qrels[qid]
        left_hits = set(left.get(qid, {})) & relevant
        right_hits = set(right.get(qid, {})) & relevant
        common = left_hits & right_hits
        left_only = left_hits - right_hits
        right_only = right_hits - left_hits
        delta = len(right_hits) - len(left_hits)

        total_relevant += len(relevant)
        total_common += len(common)
        total_left_only += len(left_only)
        total_right_only += len(right_only)
        total_left_hits += len(left_hits)
        total_right_hits += len(right_hits)

        rows.append({
            "qid": qid,
            "relevant": len(relevant),
            f"{args.left_name}_hits": len(left_hits),
            f"{args.right_name}_hits": len(right_hits),
            "common": len(common),
            f"{args.left_name}_only": len(left_only),
            f"{args.right_name}_only": len(right_only),
            "delta_right_minus_left": delta,
            f"{args.left_name}_only_docs": ";".join(sorted_docs(left_only, left.get(qid, {}))[:10]),
            f"{args.right_name}_only_docs": ";".join(sorted_docs(right_only, right.get(qid, {}))[:10]),
        })

    print(f"queries={len(qids)} k={args.k} relevant={total_relevant}")
    print(f"{args.left_name}_hits={total_left_hits} {args.right_name}_hits={total_right_hits} delta={total_right_hits - total_left_hits}")
    print(f"common={total_common} {args.left_name}_only={total_left_only} {args.right_name}_only={total_right_only}")
    if total_relevant:
        print(f"recall_{args.left_name}={total_left_hits / total_relevant:.6f} recall_{args.right_name}={total_right_hits / total_relevant:.6f}")

    worst_for_right = sorted(rows, key=lambda row: row["delta_right_minus_left"])[:args.top]
    best_for_right = sorted(rows, key=lambda row: row["delta_right_minus_left"], reverse=True)[:args.top]

    print("\nWorst right-minus-left queries:")
    for row in worst_for_right:
        print(row)

    print("\nBest right-minus-left queries:")
    for row in best_for_right:
        print(row)

    if args.out_prefix:
        output_path = Path(args.out_prefix + ".per_query.tsv")
        with output_path.open("w", encoding="utf-8", newline="") as handle:
            fieldnames = list(rows[0].keys()) if rows else ["qid"]
            writer = csv.DictWriter(handle, delimiter="\t", fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(rows)
        print(f"wrote {output_path}")


if __name__ == "__main__":
    main()
